### ISSUE TITLE
The '#target.name#-frameworks' shell scripts produce extra copies of the embedded frameworks.

### ISSUE THREAD

**Issue by Maxim-38RUS-Zabelin** [CocoaPods/CocoaPods#3361]
Integrating a third-party dynamic library into a compound workspace via Cocoapods, one may face the following issue: it's impossible to run any assembly on a real device running iOS; installed, the app crashes at launch time.

For details please refer to [this web page](https://github.com/Maxim-38RUS-Zabelin/compound-workspaces-along-with-cocoapods).

**Comment by rvangraan** [comment]
I'm having the same issue with a compound workspace that consists of two dynamic frameworks and an app:
![screen shot 2015-04-14 at 07 46 59](https://cloud.githubusercontent.com/assets/2574318/7131729/74ed42fe-e27a-11e4-80d9-6e3d63afe2c5.png)

The Podfile is something like this:

https://gist.github.com/rvangraan/c873e4a32e7f45004c8c

Even though it "worked", the build process was super slow and upon submission Apple rejected our App because it included frameworks within Frameworks. 

I opened the app bundle and saw that indeed, each of my frameworks included the same Pod frameworks again. I have no idea how to fix this.

**Comment by Maxim-38RUS-Zabelin** [comment]
@rvangraan,

Have you checked the [solution](https://github.com/Maxim-38RUS-Zabelin/compound-workspaces-along-with-cocoapods/tree/solution) branch of the repository I mentioned above? It's source code also doesn't compile really fast, but no framework duplicates are put into the archives. Please also have a look a these screenshots: [an assembly of the original workspace that manifest the issue] (http://joxi.ru/8AnXKX4ux3PMmO) and [an assembly of the updated workspace with no such an issue](http://joxi.ru/EA4NlNJSq7eVmb).

If you have any question, please do not hesitate to contact me.

**Comment by rvangraan** [comment]
Hi @Maxim-38RUS-Zabelin 

I did use your solution. It works. But after doing that, Archive started failing silently:
![screen shot 2015-04-14 at 18 04 39](https://cloud.githubusercontent.com/assets/2574318/7142482/cb2473c6-e2d0-11e4-8638-2555621005f3.png)
It produces an archive without without an Info.plist:
![screen shot 2015-04-14 at 18 06 42](https://cloud.githubusercontent.com/assets/2574318/7142512/070c9d78-e2d1-11e4-93c0-008e42183fd9.png)

It should have looked like this:
![screen shot 2015-04-14 at 18 07 19](https://cloud.githubusercontent.com/assets/2574318/7142520/1df92218-e2d1-11e4-92bb-18ba1bd35c7b.png)

The app still works in the simulator and on my phone, and all the tests passes. It is just the archiving that fails. 

The difference between the two builds are this in the Podfile:

```
# See https://github.com/CocoaPods/CocoaPods/issues/3361
post_install do |installer|
    installer.project.targets.each do |target|
        %x( if [ -f "Pods/Target Support Files/#{target.name}/#{target.name}-frameworks.sh" ]; then
           if [ "#{target.name}" != "Pods-Expend" ]; then
           echo "#!/bin/sh\necho Skipping Pods/Target Support Files/#{target.name}/#{target.name}-frameworks.sh" > "Pods/Target Support Files/#{target.name}/#{target.name}-frameworks.sh";
           fi
           fi )
    end
end
```

I've been scrutinising the build/archive logs, but I just can't see anything obviously causing the issue.

**Comment by Maxim-38RUS-Zabelin** [comment]
In fact, the difference in the Podfile should not be the only one. Just to make sure you did everything right: please check [demo-frameworks.sh](https://github.com/Maxim-38RUS-Zabelin/compound-workspaces-along-with-cocoapods/blob/solution/demo/demo/demo-frameworks.sh); have you included the last one into your build phase?

Considering the archives, to tell the truth, it's Jenkins who does me a favor and builds our apps; and we can easily install any .ipa it builds on any of our devices. I have also checked one of the .ipas: it includes [an Info.plist, a code signature and etc](http://joxi.ru/E2pvavYSeRQ4rY). To me, everything looks OK.

I guess whether the issue is caused by **Xcode** or not. As you might know Jenkins is unable to run **Xcode** and click the 'Build' button; thus it uses *xcodebuild* to get the job done. Have you tried to archive your app via *xcodebuild*?

**Comment by rvangraan** [comment]
Hi @Maxim-38RUS-Zabelin 

Thank you for that. I've added that script to the build and tried again. Same issue.

I then cloned your repo to see if it works and it resulted in exactly the same problem as I have, namely an archive with lots of stuff missing (Info.plist, etc):
https://gist.github.com/1fea8297e1f45b2a9385.git

![screen shot 2015-04-15 at 12 17 31](https://cloud.githubusercontent.com/assets/2574318/7157647/6815c944-e369-11e4-81c4-abc1b9ad8d2f.png)
![screen shot 2015-04-15 at 12 17 45](https://cloud.githubusercontent.com/assets/2574318/7157652/70bacb58-e369-11e4-813f-d36d1f584c88.png)
![screen shot 2015-04-15 at 12 18 33](https://cloud.githubusercontent.com/assets/2574318/7157665/913636e2-e369-11e4-8586-d4c2502d8890.png)

**Comment by Maxim-38RUS-Zabelin** [comment]
Have you tried to archive your app by using *xcodebuild*?

At the moment I'm not able to archive any app, but I'll look into the issue you describe ASAP. Once I get any result, I'll let you know.

**Comment by rvangraan** [comment]
Hi @Maxim-38RUS-Zabelin, Yes I tried that. But not with the last set of changes or your app yet. I will do that next. Thanks, R

**Comment by Maxim-38RUS-Zabelin** [comment]
@rvangraan,

You are welcome! Please keep me posted: it would be nice to know whether your issue is due to imperfection of the **Xcode** or not.

**Comment by rvangraan** [comment]
Hi @Maxim-38RUS-Zabelin,

I've now tried building your project with xcodebuild and it resulted in the same broken archive. I am beginning to think it is a bug in Xcode now. I'm using the latest 6.3 Release. What is strange about the archive from your project is that the swift libraries are included 3 times!
![screen shot 2015-04-15 at 14 50 11](https://cloud.githubusercontent.com/assets/2574318/7160075/c0638a4a-e37e-11e4-8d27-63b7b8dd5181.png)

**Comment by rvangraan** [comment]
@Maxim-38RUS-Zabelin,

Do you think it might be caused by #3240 ?

**Comment by Maxim-38RUS-Zabelin** [comment]
@rvangraan,

To tell the truth, I don't know, but these issues might be connected. Actually, the thing that bothers me is why our Jenkins assembles valid archives. I'm gonna check the console output (that our Jenkins generates) and compare it with one the **Xcode** produces. I'll keep you posted.

**Comment by rvangraan** [comment]
Hi,
I eventually decided to abandon the use of CocoaPods for the dynamic Frameworks in our App. It is so broken that it is not useable. 

Thanks for the help!

R

**Comment by samdmarshall** [comment]
Hi, I have read over this thread and looked at the repo you provided as an example for this and I don't think I fully understand what behavior you want to be seeing here. From what I have read, your workspace is setup to work around the issue of embedding the swift standard libraries multiple times, but results in other issues because cocoapods doesn't expect you to do this -- do I have that right?

**Comment by Maxim-38RUS-Zabelin** [comment]
Hi @samdmarshall,

You're almost right, but there are several points to clarify. The first one is it's not only the swift standard libraries that embedded multiple times - the set includes both the standard libraries and the third-party ones (the pods). The second one is CocoaPods doesn't expect me to integrate them into that compound workspaces. To overcome this issue I disable the *#target.name#-frameworks.sh* CocoaPods provides and run my own shell script (that in fact is a copy of the shell scripts CocoaPods generates) in the very end. I guess I do nothing 'illegal', do I?

**Comment by Maxim-38RUS-Zabelin** [comment]
@rvangraan,

Have you tried to use *xcrun* for archiving? This doesn't solve the issue, though it properly archives the apps. The syntax is as follows: /usr/bin/xcrun -sdk iphoneos PackageApplication -v *path_to_your_app* -o *path_to_your_ipa*.

**Comment by samdmarshall** [comment]
So I've taken another look over the workspace you provided as an example and I think I understand what you are trying to achieve with the separate Xcode projects (please correct me if i'm getting this wrong): 

> The workspace file contains a different project file for each individual framework you are building, each of these frameworks have a dependency on a library that is added to the workspace via CocoaPods. The separate project files exist to separate the specific frameworks you have and make them easier to manage and more modular.

This is all allowed and what you are seeing is what I would interpret as expected behavior on behalf of Xcode. From how your target dependencies are setup, you are linking the same library (core.framework) into the Demo target, as well as into the pad.framework and phone.framework. Xcode treated all of these are separate builds; independent of what any other target depends on. This means that regardless if you use CocoaPods or not in your current setup - you would have to manually resolve this shared dependency so it only gets added once.

Same principle applies to the multiple swift standard libraries that get added. Seeing that you linked my write-up of the current problem of `EMBEDDED_CONTENT_CONTAINS_SWIFT` in the repo readme, i won't summarize that again and instead jump right to saying that the multiple instances of the swift libraries being packaged would also be a problem that you would have to resolve manually outside of using CocoaPods. 

As I am understanding this, your issue is two-fold: First, being that you have separate targets that use the same set of dependencies, then having that be exacerbated by the scripts used by CocoaPods to mitigate the "bugs" in Apple's tooling. Your suggestion of being able stop CocoaPods creating the workaround scripts would work and I can see why removing CocoaPods would be a path of least resistance in resolving this issue for you. That would still leave you with the problem of having your frameworks share dependencies and cause your application's size to ballon because they would be included multiple times. 

I would propose that you instead change how your frameworks are configured so that they don't share the same dependencies. This could be anything from using weak-linking on the libraries included from CocoaPods -- to drastically reconfiguring how your frameworks are setup so that they don't depend on the same libraries.

I don't think that working with multiple projects in a workspace with CocoaPods is fundamentally flawed or not possible -- it definitely raises the level of difficulty when dealing with dependencies and will highly even the smallest potential conflict in management. I think there should be a happy medium between creating workarounds to this and a complete overhaul, hopefully one of the CocoaPods core team can chime in and suggest something for this.

**Comment by rvangraan** [comment]
Hi @samdmarshall 
> I would propose that you instead change how your frameworks are configured so that they don't share the same dependencies. This could be anything from using weak-linking on the libraries included from CocoaPods -- to drastically reconfiguring how your frameworks are setup so that they don't depend on the same libraries.
In my case the dependencies weren't shared between the targets. For example, one of the embedded frameworks referenced a Pod and the main target did not. This was enough to break the Archiving completely. I used @Maxim-38RUS-Zabelin's workaround on my build, but there was no way for me to get the Archiving to generate a good artifact.

**Comment by samdmarshall** [comment]
Could you upload an example so I have something to use to reproduce this issue with? (If possible sans swift because I am running 10.9.5 and cannot build the latest swift version) I would like to investigate this and resolve it as I might have overlooked these issues you are having in my original analysis and research that was to help CocoaPods implement the scripts to work around the duplicate library issue.

**Comment by Maxim-38RUS-Zabelin** [comment]
@samdmarshall,

I agree with you, though I'd like to clarify one thing: I didn't really want to make my frameworks share the same pods. I did so because I had seen no other option to build the entire workspace. The original idea was to integrate the *core* with a set of pods and use it as a foundation for the *phone* and *pad* frameworks. In other words, I wanted the *core* to aggregate an API to communicate with a remote web-service while the *phone* and *pad* were expected to share this core and provide a unique GUI for each interface idiom. Originally, only the *core* was integrated with the Alamofire via CocoaPods (to be able to communicate with a remote web-service); unfortunately, every time I tried to import the *core* into either the *phone* or *pad*, the compiler claimed "missing required module 'Alamofire'".

Could you tell me if there is any way to make a nested dynamic framework (in this case this framework is Alamofire) visible from outside another dynamic framework (in this case it is the *core*)? I would appreciate any suggestion!

**Comment by samdmarshall** [comment]
The organizational decision of creating separate framework for the UI on each the iPad and iPhone seems strange to me unless you are also intending to release these to be consumed by other people. If these are just for your own app, then I would strongly recommend creating a single UI framework that would be agnostic enough to work on both platforms. If you really need to integrate this way then it might be possible to create your own podspecs for the _core_, _pad_, and _phone_ frameworks to make them integrate cleanly with the rest of the Pods you are using. 

While the project organization that you are talking about here is possible, I don't think I would recommend organizing it in that manner. You are going to be dealing with duplicated libraries or dealing with weak-linking the libraries and setup search paths so that they can be resolved to build and not linked with directly (and thus included more than once).

If this could be updated with an example repo of your ideal project organization that i could build and play with -- I should be able to have a look at what could be done about this over the weekend. (If it could be an ObjC project that would be super)

**Comment by Maxim-38RUS-Zabelin** [comment]
Hi @rvangraan, @samdmarshall,

It seems I've finally made it: please have a look at [one more solution](https://github.com/Maxim-38RUS-Zabelin/compound-workspaces-along-with-cocoapods/tree/another_solution) to the originally described issue. This time you should be able to archive the demo app without any problem. And it would be nice to have a feedback from you. Thank you in advance!

**Comment by samdmarshall** [comment]
Hi, as mentioned I am not able to run a version of Xcode that supports Swift 1.2, so the sample you have created won't build at all for me. While i could work around this issue myself i also don't want to accidentally add any work-arounds or fixes of my own in the process that would change the behavior you have setup already. Also I am not 100% clear as to how your new project organization works now you have removed the 3 other projects and still depend on their build products.

**Comment by neonichu** [comment]
This should be fixed by #3550

**Comment by Maxim-38RUS-Zabelin** [comment]
To tell the truth, I don't see any changes. I've updated Cocoapods to v.0.38.0.beta.1 and the resulting archive still has duplicates (http://joxi.ru/Vm6yNygc9jjZrZ). Did I miss something?
